File: sorting_algorithms/merge_sort/python/merge_sortv4.py
import time

def BottomUpMerge(A, start, mid, end, B):
    i = start 
    j = mid

    for k in range(start, end):
        if (i < mid and (j >= end or A[i] <= A[j])):
            B[k] = A[i]
            i += 1
        else:
            B[k] = A[j]
            j += 1

def BottomUpMergeSort(A, B, n):
    width = 1
    while(width < n):
        for i in range(0, n, 2 * width):
            BottomUpMerge(A, i, min(i + width, n), min(i + 2 * width, n), B)

        A = [k for k in B]
        width *= 2

    return A

A = [5, 2, 4, 6, 1, 3]
B = [None] * len(A)

A = BottomUpMergeSort(A, B, len(A))

print(list(A))

File: sorting_algorithms/merge_sort/C/merge_sortv4.c
#include <stdio.h>
#include <math.h>

void BottomUpMerge(int A[], int start, int mid, int end, int B[]) {
    int i = start;
    int j = mid;

    for (int k = start; k < end; k++) {
        if (i < mid && (j >= end || A[i] <= A[j])) 
            B[k] = A[i++];
        else
            B[k] = A[j++];
    }
}

void BottomUpMergeSort(int A[], int B[], int n) {
    for (int width = 1; width < n; width = 2 * width) {
        for (int i = 0; i < n; i = i + 2 * width) {
            BottomUpMerge(A, i, fmin(i + width, n), fmin(i + 2 * width, n), B);
        }

        for (int k = 0; k < n; k++)
            A[k] = B[k];
    }
}

int main() {
    int A[6] = {5, 2, 4, 6, 1, 3};
    int B[6];

    BottomUpMergeSort(A, B, 6);

    for (int i = 0; i < 6; i++)
        printf("%d, ", A[i]);
    printf("\n");

    return 0;
}